Celery runner

Kirill888 opened the pull request:
# Overview
New executor that uses Celery (with Redis as broker and data backend).

This provides an alternative to current setup (`dask.distributed`). Problem with using `dask.distributed` is that it requires that tasks are idempotent, since it will sometimes schedule the same task in parallel on different nodes. With many tasks doing I/O this creates problems.

Celery in comparison has a much simpler execution model, and doesn't have same constraints.

## Redis backend

Celery supports a number of backends, of them two are fully supported: RabbitMQ and Redis. I have picked Redis as it is the simplest to get running without root access (NCI environment)

## data_task_options

Adding `celery` option to `--executor` command line, same `host:port` argument is used. Celery executor will connect to Redis instance at a given address, if address is localhost and Redis is not running, it will be launched for the duration of the execution. Workers don't get launched however, so in most cases the app will stall until workers are added to the processing pool (see datacube-worker)

`$HOME/.datacube-redis` contains redis password, if this file doesn't exist it will be created with a randomly generated password when launching Redis server.

Also adding executor alias `dask` to be the same as `distributed`. However now that we have 2 distributed backends we should probably favour `dask` as a name for `dask.distributed` backend.

## datacube-worker

New app `datacube-worker` was added to support launching workers in either `celery` or `dask` mode. It accepts the same `--executor` option as the task app.

Kirill888 committed:
Adding __repr__ to some executor classes

Convenience method for interactive development/debugging.

Kirill888 committed:
Sample task_app

Sample app that uses executor to run tasks in parallel

Kirill888 committed:
Adding celery based executor

Currently one can not set redis address on a command line, but it can be set via
environment variable `REDIS=redis://<host>:>port>/<db>`

Kirill888 committed:
Allow command line configuration or celery

Redis server IP/port can be configured from command line now.

Kirill888 committed:
pep8 fixes for dummy task app

Kirill888 committed:
Adding celery+redis to travis env

Kirill888 committed:
Celery runner launches redis when needed

Also shutdowns celery workers and redis when done, including redis-server as
part of the environment.

Kirill888 committed:
Fixing num_tasks argument handling

When not supplied on a command line should be use one from the config, instead
it was setting to None.

Kirill888 committed:
Fixing code check problems

Kirill888 committed:
Python 2.6 doesn't have ConnectionError

changed to IOError

Kirill888 committed:
Adding worker launcher python app

`datacube-worker` accepts the same `--executor` option as task apps and launches
appropriate worker task: `celery::redis` or `dask::distributed` based worker.

Kirill888 committed:
Adding test script for celery_runner

This one launches worker and app on a local machine only

Kirill888 committed:
Changed redis config

Allowing remote connections to redis without password for testing, for now

Kirill888 committed:
Adding dask as an alias for `distributed`

when launching `datacube-worker` allow `dask` as an alias for `distributed`

Kirill888 committed:
Moved dummy task app

Also adding shell script for launching on pbs

Kirill888 committed:
minor style correction

2 spaces between functions

Kirill888 committed:
Set executable flags on launch_pbs shell script

Kirill888 committed:
Adding some util functions

- gen_password
- slurp content of a file
- write_user_secret_file function to create files that are readable by user only

Kirill888 committed:
Adding password protection to celery::redis

By default now redis server will be launched with a randomly generated password
that will be stored in user home directory in a text file only user has
permissions to read. Celery workers load password from that file, and so need to
have access to the home directory (ok on NCI) but also need to be launched after
the file was created.

Password is generated once and then re-used unless user manually deletes the
file where it is stored, `.datacube-redis` at the moment.

Kirill888 committed:
Fixing pylint complaints

in this case IOError should be enough

Kirill888 committed:
Fixing pylint complaints

removed `params` options, just use `key=value` syntax. For redis config options
that have `-` in the name replace it with `_` to fit python syntax constraints.

Kirill888 committed:
Celery version constraint

also updated setup.py to include celery+redis dependency.

Kirill888 committed:
Using pip for latest version of celery

conda doesn't have celery 4 yet.

omad committed:
Merge branch 'develop' into celery_runner

Kirill888 committed:
Adding tests for new functions in util

- write_user_secret_file
- slurp
- gen_password

Fixing an error in `write_user_secret_file`, open mode was missing expected
truncate flag, hence appending to existing file. Making open mode binary/text
settable from the caller in `slurp` and `write...` functions

Kirill888 committed:
bumped minimum password size while testing

Code checks that two random passwords are not the same, with only 1 byte entropy
that sometimes is not the case.

Kirill888 committed:
Adding tests for redis part of celery runner

Unifying treatment of password argument, empty string always implies load
password from file, or generate and store depending on context.

Kirill888 committed:
More test for celery executor

So far testing task submission only, no worker is started in the tests.

Kirill888 committed:
End to end testing of celeryexecutor

Kirill888 committed:
Merge remote-tracking branch 'origin' into celery_runner

Kirill888 committed:
Moved pickled function into test scope

Kirill888 committed:
Adding some debug to circleci tests

Kirill888 committed:
Making redis tests optional

If no redis-server is found on the test machine, then don't run tests that
require redis-server

Kirill888 committed:
Moved common part of pbs launching script into pbs_helper.sh

Kirill888 committed:
Fixing up pbs_launch.sh script

Need to activate environment before you can source helper shell script

Kirill888 committed:
Merge branch 'develop' of github.com:opendatacube/datacube-core into celery_runner

andrewdhicks commented on datacube/_celery_runner.py:
```diff
@@ -0,0 +1,259 @@
+                # If no change detected sleep for a bit
+                # TODO: this is sub-optimal, not sure what other options are
+                #       though?
+                sleep(0.1)
```
It is a shame that `celery.ResultSet` doesn't have a `collect()` function like `AsyncResult`  http://docs.celeryproject.org/en/latest/reference/celery.result.html#celery.result.AsyncResult.collect

andrewdhicks commented on datacube/ui/click.py:
```diff
@@ -206,9 +206,12 @@ def parse_endpoint(value):
+    'celery': lambda addr: mk_celery_executor(*parse_endpoint(addr))
 }
 
+EXECUTOR_TYPES['dask'] = EXECUTOR_TYPES['distributed']  # Add alias "dask" for distributed
```
This could make things more confusing, as dask also has synchronous, multi-threaded and multi-process schedulers/executors.

Kirill888 replied:
Ok, but 'distributed' is too generic also as celery is also "distributed" across machines, and 'dask[.-_]distributed' is hard to type and to remember separator token.

andrewdhicks replied:
Good point!

andrewdhicks commented on datacube/ui/task_app.py:
```diff
@@ -263,6 +263,14 @@ def add_dataset_to_db(index, datasets):
+def wrap_task(f, *args, **kwargs):
+    'turn function `f(task, *args, **kwargs)` into `g(task)` in pickle-able fashion'
+    return functools.partial(_wrap_impl, f, args, kwargs)
+
```
We should aim for PEP 257 with docstrings - Triple quotes are used even though the string fits on one line. This makes it easy to later expand it.

Kirill888 committed:
Fixing formatting for doc strings

addressing code review, also deleted unused import

Kirill888 committed:
Fix for newer version of netCDF4 python wrapper

chartostring now returns decoded string always, so numpy.char.decode step is no
longer needed. Code should still work with older lib.

Kirill888 committed:
Fixing netcdf4 to be less than 1.2.8

Two things changed in 1.2.8
- chartostring
- cython version dependency

cython thing somehow broke compliance checkers, it can not load any checkers as
they fail with

```
DistributionNotFound: The 'cython>=0.19' distribution was not found and is required by netCDF4
```
in `load_all_available_checkers`, so no checkers are loaded which then asserts
in `run` method.

Kirill888 committed:
String handling for netcdf variable

Adding new function `netcdf_extract_string` that takes care of possible ways
strings can be stored in netcdf. This fixes test failures when using 1.3.8
version that switched to returning unicode from `chartostring` method.

Kirill888 committed:
Forcing installation of cython to support netcdf4 1.2.8

cython is a compulsory dependency of netcdf4 starting from 1.2.8, but since it's
a build-time dependency conda doesn't install it, netdcf4 is compiled
externally. Lack of cython doesn't break netcdf4 library, but it does break
compliance checker, because it uses `pkg_resource` lib to check if dependencies
for plugins are installed and `pkg_resource` reports that cython is needed to
import `netcdf4` which isn't true.

woodcockr commented:
Hi Folks,
I've been watching this branch with interest. Looks good, I like the switching of different execution models. When you are ready would love a chat/doc about how this design all comes together and where you might be heading next to help with some of our future contributions.
Thanks for the effort!

Kirill888 commented:
Hi Rob,

thank you for feedback. In a very short term we are hoping to have this branch merged in it's current state. Changes are backwards compatible with processing apps we currently have and we will likely leave them be as they are for now. Longer term I think we can all agree that writing "task apps" can be a lot simpler, and you can consider work in this branch as an exploration of pain points. I do not claim to address any of them just yet.

Ideally I'd like us to move away from writing custom launch scripts with hard-coded environment setup in bash and using `qsub` on a command-line, I think all this logic can be done in python and re-used by all processing apps with an `import` and a decorator.

I want to be able to write processing app that can be easily tested locally, and then without writing any scripts deployed on NCI, or any other compute provider we might support in the future.

```
  source activate my-env
  # run locally to test
  ./my-processing-app.py --my-app-option1 --my-option2=small-dataset

  # run on a raijin using qsub under the hood
  ./my-processing-app.py --pbs="num-nodes=10,time=1h" --my-app-option1 --my-option2=large-dataset

  # run on a AWS using dockers or whatever 
  ./my-processing-app.py --aws="num-nodes=10,region=oregon" --my-app-option1 --my-option2=large-dataset
```

I will be away in June and everyone else have their plate full with higher priority tasks as it is, so I do not expect any more work to be done in that direction for a while.

Kirill888 committed:
Adding more executor tests

also removed netcdf4 peg in `setup.py`, should have done it in the last commit.

woodcockr commented:
Great. That is exactly we wanted to achieve with the Execution Engine - write your app, run it on any of the target compute resource environments. preferably with it figuring out the best way to divide up the tasking.

I've made a note in our planning to link up with this once we clear the current development tasks.

Kirill888 committed:
More executor tests

somewhat higher coverage for executor classes.

Kirill888 committed:
Telling conda to cleanup before running tests

running out of disk on travis test machines, see if this helps

Kirill888 committed:
Switching parallel executor to use cloudpickle unless disabled

To be feature compatible with dask, celery and serial executor. Now you can
submit lambda's and inner functions not just top-level functions.

Kirill888 committed:
Making sure files are closed

Fixing `ResourceWarning` (file handle leak) in `read_documents`

Kirill888 committed:
Moved conda clean command

Still running out of disk on travis, conda clean should happen after setting up
agdc environment.

Kirill888 committed:
`file` is no good name on py27, `file`->`handle`

Fixing pylint complaints, also minor flake8 formatting corrections.

omad reviewed (approved):
Looks good enough to merge. Since Kirril is away for the rest of June I'm not going to let this wait any longer.
